proposals/0001-lhls.md

```diff
@@ -0,0 +1,313 @@
+- Feature Name: Low-Latency Streaming
+- Start Date: 2018-08-18
+- RFC PR: https://github.com/video-dev/hlsjs-rfcs/pull/1
+- Hls.js Issue: (leave this empty)
+- Authors: John Bartos, Tom Boshoven
+- Signatories: JW Player, (your name here!)
+
+# Summary
+[summary]: #summary
+
+Low-latency streaming is becoming an increasingly desired feature for live events, and is typically defined as a delay of two seconds or less from point of capture to playback (glass-to-glass). However, the current HLS specification precludes this possibility - within the HLS guidelines, the best attempts have achieved about four seconds glass-to-glass, with average implementations typically beyond thirty seconds. This RFC proposes modifications to the HLS specification ("HTTP Live Streaming 2nd Edition" specification (IETF RFC 8216, draft 03)") which aim to reduce the glass-to-glass latency of a live HLS stream to two seconds or below. The scope of these changes are centered around a new "prefetch" segment; it's advertising, delivery, and interpretation within the client.
```

Comment:
This is kind of a moot point because I think 2 seconds is a good goal for this project, but 2 seconds is probably the most aggressive definition of "low latency" I've seen. [Wowza](https://static1.squarespace.com/static/5955fc221b631b7dd41ede60/t/5ab4cc930e2e72bd834f7a46/1521798310775/Ultra_Low_Latency_Streaming_for_When_Every_Second_Counts_Wowza.png) pegs it at 1-5s and @wilaw [has it at 4-10s](https://www.twitch.tv/demuxed/clip/TastyJoyousPonyPRChase), with 2 seconds being closer to "Ultra low latency". Not sure what I expect you to do with that info but thought it was worth pointing out in case there's opportunity for industry consistency.

Reply:
Agreed. I've been working with Will on the new latency ranges definition that you have seen in his Demuxed presentation. We defined low latency by what you can achieve with 1s and 2s segments of regular HLS/DASH (meaning : 4 to 10 seconds latency) and ultra low latency as what you can achieve with chunked CMAF (meaning : between 1 and 4 seconds). We used this technology criteria as the previous latency ranges definition by Wowza & Streaming Media was mixing technology and use case requirements criterias.
![latency_ranges-v3](https://user-images.githubusercontent.com/38870794/51024086-df717100-1588-11e9-8d3b-eab88837a83f.png)

So we maybe should say here: 'Low-latency and Ultra low-latency streaming are becoming  increasingly desired features for live events, and are typically defined as a delay of 4 to 10 seconds for low latency and 1 to 4 seconds for ultra low latency, from point of capture to playback (glass-to-glass)'

Reply:
Yeah agreed. Thanks for keeping me up to date with Will's work - I want to be aligned with whatever he's doing where possible. Has he been factoring the Streamline project into his definitions? The LHLS fork of Exoplayer is getting 1.1s latency which is pretty crazy. 

>'Low-latency and Ultra low-latency streaming are becoming increasingly desired features for live events, and are typically defined as a delay of 4 to 10 seconds for low latency and 1 to 4 seconds for ultra low latency, from point of capture to playback (glass-to-glass)'

Sounds good to me!

As of writing this, Hls.js by default will have between 1 and 2 segment's duration of client-side latency. The plan is to start the stream at the last complete segment, and begin buffering prefetch from there. But this will be open to configuration.

Reply:
>We defined low latency by what you can achieve with 1s and 2s segments of regular HLS/DASH (meaning : 4 to 10 seconds latency) and ultra low latency as what you can achieve with chunked CMAF (meaning : between 1 and 4 seconds). 

So by this definition we're building an ultra low latency player.

Reply:
By default yes, but the player configuration shall allow a higher target latency. Which leads me to another close consideration : in DASH we are heading towards setting the target latency on the manifest level, and not on the player configuration level. Would it be interesting for LHLS to discuss a similar approach, like #EXT-X-TARGETLATENCY: 2500 (if we measure in milliseconds) ?

Reply:
@nicoweilelemental That's an interesting idea!!! There are lot of advantages in being able to configure the playout latency at the manifest itself. If we going ahead with this, I propose a small modification though. Instead of setting a target latency, I suggest we set a target buffer size. Since the encoder's latency is not in player's control this definition could be little-bit misleading. Instead we could ask the player to maintain a specific target buffer size with the condition that it continuously loads data to be as close to the live edge as possible. Something like #EXT-X-TARGETBUFFERSIZE: 2500 .

Reply:
You're right @jkarthic-akamai it's hard for the player to determine what's the actual E2E latency. The only way would be to make at least one #EXT-X-PROGRAM-DATE-TIME insertion mandatory per child playlist. In dash we recommend putting the Producer Reference Time in the prft mp4 box, which the player can parse to get the actual timecode of a segment: it could replace #EXT-X-PROGRAM-DATE-TIME for latency measurement.
Setting the #EXT-X-TARGETBUFFERSIZE instead will roughly fill the same purpose while relaxing the dependency to absolute time.